<p style="font-family: Arial; font-size:3em;color:black;"> Session 4 - Lab Exercise 7</p>

We will use the titanic dataset from Kaggle (https://www.kaggle.com/). This is a well-known dataset and we will use it for classification- if the passenger survived or passed away.

In [145]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

In [146]:
training_DF = pd.read_csv('titanic_dataset_GBC.csv')

In [147]:
training_DF['Age'].fillna(training_DF['Age'].mean(), inplace=True)

In [148]:
training_DF.drop('Cabin',axis=1,inplace=True)

In [149]:
training_DF.dropna(inplace=True)

In [150]:
sex = pd.get_dummies(training_DF['Sex'])
embark = pd.get_dummies(training_DF['Embarked'])

In [151]:
training_DF.drop(['Sex','Embarked','Name','Ticket'],axis=1,inplace=True)

In [152]:
training_DF = pd.concat([training_DF,sex,embark],axis=1)

In [153]:
training_DF.drop(['female','C'],axis=1,inplace=True)

In [154]:
from sklearn.model_selection import train_test_split

In [155]:
X_train, X_test, y_train, y_test = train_test_split(training_DF.drop('Survived',axis=1), 
                                                    training_DF['Survived'], test_size=0.30, 
                                                    random_state=101)

In [156]:
from sklearn.linear_model import SGDClassifier
from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler()
X_train = scaler.fit_transform(X_train)



In [157]:
# let's explore SGDClassifier parameters


clf=SGDClassifier(
    loss='hinge',
    penalty='l2',
    alpha=0.0001,
    l1_ratio=0.15,
    fit_intercept=True,
    max_iter=1000,
    tol=0.001,
    shuffle=True,
    verbose=0,
    epsilon=0.1,
    n_jobs=None,
    random_state=None,
    learning_rate='optimal',
    eta0=0.0,
    power_t=0.5,
    validation_fraction=0.1,
    n_iter_no_change=5,
    class_weight=None,
    warm_start=False,
    average=False,
)

# Let's form SGD models with variation in paameters loss and alpha
    # loss: 'hinge', 'log', and 'modified_huber'
    # alpha: 0.0001 and 0.001
    # explain and dicuss your findings

In [158]:
clf.fit(X_train,y_train)

SGDClassifier()

In [159]:
clf.get_params()

{'alpha': 0.0001,
 'average': False,
 'class_weight': None,
 'early_stopping': False,
 'epsilon': 0.1,
 'eta0': 0.0,
 'fit_intercept': True,
 'l1_ratio': 0.15,
 'learning_rate': 'optimal',
 'loss': 'hinge',
 'max_iter': 1000,
 'n_iter_no_change': 5,
 'n_jobs': None,
 'penalty': 'l2',
 'power_t': 0.5,
 'random_state': None,
 'shuffle': True,
 'tol': 0.001,
 'validation_fraction': 0.1,
 'verbose': 0,
 'warm_start': False}

In [160]:
prediction = clf.predict(X_test)    #Case 1
print("prediction:",prediction)


from sklearn.metrics import accuracy_score
print("accuracy:",accuracy_score(y_test,prediction))

from sklearn.metrics import confusion_matrix
confusion_matrix(y_test,prediction)


prediction: [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0]
accuracy: 0.6067415730337079


c:\Users\chunl\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\base.py:457: UserWarning: X has feature names, but SGDClassifier was fitted without feature names
  warnings.warn(


array([[162,   1],
       [104,   0]], dtype=int64)

In [161]:
clf=SGDClassifier(   # 2log
    loss='log_loss',
    penalty='l2',
    alpha=0.0001,
    l1_ratio=0.15,
    fit_intercept=True,
    max_iter=1000,
    tol=0.001,
    shuffle=True,
    verbose=0,
    epsilon=0.1,
    n_jobs=None,
    random_state=None,
    learning_rate='optimal',
    eta0=0.0,
    power_t=0.5,
    validation_fraction=0.1,
    n_iter_no_change=5,
    class_weight=None,
    warm_start=False,
    average=False,
)
clf.fit(X_train,y_train)
clf.get_params()

{'alpha': 0.0001,
 'average': False,
 'class_weight': None,
 'early_stopping': False,
 'epsilon': 0.1,
 'eta0': 0.0,
 'fit_intercept': True,
 'l1_ratio': 0.15,
 'learning_rate': 'optimal',
 'loss': 'log_loss',
 'max_iter': 1000,
 'n_iter_no_change': 5,
 'n_jobs': None,
 'penalty': 'l2',
 'power_t': 0.5,
 'random_state': None,
 'shuffle': True,
 'tol': 0.001,
 'validation_fraction': 0.1,
 'verbose': 0,
 'warm_start': False}

In [162]:
prediction2 = clf.predict(X_test)    #Case 2
print("prediction:",prediction2)


from sklearn.metrics import accuracy_score
print("accuracy:",accuracy_score(y_test,prediction2))

from sklearn.metrics import confusion_matrix
confusion_matrix(y_test,prediction2)

prediction: [1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1
 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 1 0 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1]
accuracy: 0.3895131086142322


c:\Users\chunl\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\base.py:457: UserWarning: X has feature names, but SGDClassifier was fitted without feature names
  warnings.warn(


array([[  4, 159],
       [  4, 100]], dtype=int64)

In [163]:
clf=SGDClassifier(   # huber
    loss='huber',
    penalty='l2',
    alpha=0.0001,
    l1_ratio=0.15,
    fit_intercept=True,
    max_iter=1000,
    tol=0.001,
    shuffle=True,
    verbose=0,
    epsilon=0.1,
    n_jobs=None,
    random_state=None,
    learning_rate='optimal',
    eta0=0.0,
    power_t=0.5,
    validation_fraction=0.1,
    n_iter_no_change=5,
    class_weight=None,
    warm_start=False,
    average=False,
)
clf.fit(X_train,y_train)
clf.get_params()

{'alpha': 0.0001,
 'average': False,
 'class_weight': None,
 'early_stopping': False,
 'epsilon': 0.1,
 'eta0': 0.0,
 'fit_intercept': True,
 'l1_ratio': 0.15,
 'learning_rate': 'optimal',
 'loss': 'huber',
 'max_iter': 1000,
 'n_iter_no_change': 5,
 'n_jobs': None,
 'penalty': 'l2',
 'power_t': 0.5,
 'random_state': None,
 'shuffle': True,
 'tol': 0.001,
 'validation_fraction': 0.1,
 'verbose': 0,
 'warm_start': False}

In [164]:
prediction3 = clf.predict(X_test)    #Case 3
print("prediction:",prediction3)


from sklearn.metrics import accuracy_score
print("accuracy:",accuracy_score(y_test,prediction3))

from sklearn.metrics import confusion_matrix
confusion_matrix(y_test,prediction3)

prediction: [1 1 1 0 1 0 1 1 1 0 1 1 1 0 1 1 0 0 0 0 0 1 0 0 1 1 0 1 0 0 0 1 0 0 1 1 0
 1 0 1 1 1 1 0 1 1 1 1 0 1 1 1 0 1 1 0 0 1 1 0 0 1 1 1 1 0 1 1 0 0 0 0 1 1
 1 1 0 1 0 0 0 1 1 0 0 1 1 0 1 1 1 0 0 0 1 0 1 0 0 1 1 1 1 1 0 0 1 1 0 0 0
 1 1 1 0 1 1 1 0 0 1 0 1 0 0 1 1 0 0 0 0 1 1 1 0 1 0 1 1 1 0 0 1 1 1 0 1 1
 0 0 1 0 1 1 0 0 0 1 1 0 1 0 1 1 0 0 0 1 1 0 1 0 0 0 0 1 0 0 1 0 0 0 1 1 0
 1 1 0 0 0 0 0 0 1 1 0 0 0 1 0 1 1 1 1 1 1 1 0 1 1 0 1 1 0 0 1 1 1 1 0 1 1
 0 0 0 1 1 1 0 1 1 1 1 1 1 0 0 1 0 1 1 1 0 1 1 1 1 0 0 1 1 1 0 1 1 1 0 0 0
 0 0 1 1 1 1 1 1]
accuracy: 0.48314606741573035


c:\Users\chunl\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\base.py:457: UserWarning: X has feature names, but SGDClassifier was fitted without feature names
  warnings.warn(


array([[71, 92],
       [46, 58]], dtype=int64)

In [165]:
clf=SGDClassifier(   # huber alpha 0.001
    loss='huber',
    penalty='l2',
    alpha=0.001,
    l1_ratio=0.15,
    fit_intercept=True,
    max_iter=1000,
    tol=0.001,
    shuffle=True,
    verbose=0,
    epsilon=0.1,
    n_jobs=None,
    random_state=None,
    learning_rate='optimal',
    eta0=0.0,
    power_t=0.5,
    validation_fraction=0.1,
    n_iter_no_change=5,
    class_weight=None,
    warm_start=False,
    average=False,
)
clf.fit(X_train,y_train)
clf.get_params()

{'alpha': 0.001,
 'average': False,
 'class_weight': None,
 'early_stopping': False,
 'epsilon': 0.1,
 'eta0': 0.0,
 'fit_intercept': True,
 'l1_ratio': 0.15,
 'learning_rate': 'optimal',
 'loss': 'huber',
 'max_iter': 1000,
 'n_iter_no_change': 5,
 'n_jobs': None,
 'penalty': 'l2',
 'power_t': 0.5,
 'random_state': None,
 'shuffle': True,
 'tol': 0.001,
 'validation_fraction': 0.1,
 'verbose': 0,
 'warm_start': False}

In [166]:
prediction4 = clf.predict(X_test)    #Case 4
print("prediction:",prediction4)


from sklearn.metrics import accuracy_score
print("accuracy:",accuracy_score(y_test,prediction4))

from sklearn.metrics import confusion_matrix
confusion_matrix(y_test,prediction4)

clf.score(X_train,y_train)
clf.score(X_test,y_test)

prediction: [1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 1 0 1 1 1
 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 0 1 1
 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 0 0
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 1 0 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1
 1 1 1 1 1 1 1 1]
accuracy: 0.4157303370786517


c:\Users\chunl\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\base.py:457: UserWarning: X has feature names, but SGDClassifier was fitted without feature names
  warnings.warn(
c:\Users\chunl\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\base.py:457: UserWarning: X has feature names, but SGDClassifier was fitted without feature names
  warnings.warn(


0.4157303370786517

In [167]:
clf=SGDClassifier(   # log & 0.001
    loss='log_loss',
    penalty='l2',
    alpha=0.001,
    l1_ratio=0.15,
    fit_intercept=True,
    max_iter=1000,
    tol=0.001,
    shuffle=True,
    verbose=0,
    epsilon=0.1,
    n_jobs=None,
    random_state=None,
    learning_rate='optimal',
    eta0=0.0,
    power_t=0.5,
    validation_fraction=0.1,
    n_iter_no_change=5,
    class_weight=None,
    warm_start=False,
    average=False,
)
clf.fit(X_train,y_train)
clf.get_params()

{'alpha': 0.001,
 'average': False,
 'class_weight': None,
 'early_stopping': False,
 'epsilon': 0.1,
 'eta0': 0.0,
 'fit_intercept': True,
 'l1_ratio': 0.15,
 'learning_rate': 'optimal',
 'loss': 'log_loss',
 'max_iter': 1000,
 'n_iter_no_change': 5,
 'n_jobs': None,
 'penalty': 'l2',
 'power_t': 0.5,
 'random_state': None,
 'shuffle': True,
 'tol': 0.001,
 'validation_fraction': 0.1,
 'verbose': 0,
 'warm_start': False}

In [168]:
prediction5 = clf.predict(X_test)    #Case 5
print("prediction:",prediction5)


from sklearn.metrics import accuracy_score
print("accuracy:",accuracy_score(y_test,prediction5))

from sklearn.metrics import confusion_matrix
confusion_matrix(y_test,prediction5)

prediction: [0 0 1 1 1 0 1 0 0 0 1 1 1 0 1 1 1 0 0 0 0 0 0 0 1 1 0 1 0 0 0 0 0 0 1 1 0
 0 0 0 1 1 0 0 1 0 0 1 0 1 0 1 0 1 0 0 0 0 1 0 0 1 1 1 1 0 0 1 0 0 1 0 0 1
 0 0 0 1 0 0 0 1 1 0 0 0 1 0 1 1 0 0 0 0 1 0 1 0 0 0 1 1 1 1 0 0 1 0 0 0 0
 1 1 0 0 1 1 1 0 0 1 0 1 0 0 1 1 0 0 1 0 1 1 1 0 0 0 0 1 0 0 0 0 1 1 0 1 1
 0 0 1 0 1 1 0 0 0 0 1 0 1 0 1 1 0 1 0 0 1 0 0 0 0 0 0 1 0 0 1 0 0 0 0 1 0
 0 1 0 0 0 0 1 0 1 0 0 0 0 0 1 1 0 1 0 1 1 0 0 0 1 0 1 1 0 0 1 1 0 1 0 1 0
 0 0 0 1 0 1 0 1 1 1 0 0 1 1 0 0 0 1 1 1 0 1 1 0 0 0 0 1 0 1 0 1 1 0 0 0 0
 0 0 1 1 1 1 1 1]
accuracy: 0.5955056179775281


c:\Users\chunl\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\base.py:457: UserWarning: X has feature names, but SGDClassifier was fitted without feature names
  warnings.warn(


array([[106,  57],
       [ 51,  53]], dtype=int64)

In [169]:
clf=SGDClassifier(   # hinge & 0.001
    loss='hinge',
    penalty='l2',
    alpha=0.001,
    l1_ratio=0.15,
    fit_intercept=True,
    max_iter=1000,
    tol=0.001,
    shuffle=True,
    verbose=0,
    epsilon=0.1,
    n_jobs=None,
    random_state=None,
    learning_rate='optimal',
    eta0=0.0,
    power_t=0.5,
    validation_fraction=0.1,
    n_iter_no_change=5,
    class_weight=None,
    warm_start=False,
    average=False,
)
clf.fit(X_train,y_train)
clf.get_params()

{'alpha': 0.001,
 'average': False,
 'class_weight': None,
 'early_stopping': False,
 'epsilon': 0.1,
 'eta0': 0.0,
 'fit_intercept': True,
 'l1_ratio': 0.15,
 'learning_rate': 'optimal',
 'loss': 'hinge',
 'max_iter': 1000,
 'n_iter_no_change': 5,
 'n_jobs': None,
 'penalty': 'l2',
 'power_t': 0.5,
 'random_state': None,
 'shuffle': True,
 'tol': 0.001,
 'validation_fraction': 0.1,
 'verbose': 0,
 'warm_start': False}

In [170]:
prediction6 = clf.predict(X_test)    #Case 6
print("prediction:",prediction6)


from sklearn.metrics import accuracy_score
print("accuracy:",accuracy_score(y_test,prediction6))

from sklearn.metrics import confusion_matrix
confusion_matrix(y_test,prediction6)

prediction: [1 1 1 1 1 0 1 1 1 0 1 1 1 0 1 1 1 1 0 1 0 1 0 0 1 1 0 1 0 1 0 1 1 0 1 1 0
 1 0 1 1 1 1 0 1 1 1 1 0 1 1 1 0 1 1 0 1 1 1 0 0 1 1 1 1 0 1 1 0 1 1 0 1 1
 1 1 0 1 0 0 0 1 1 0 0 1 1 0 1 1 1 0 1 0 1 1 1 0 0 1 1 1 1 1 0 0 1 1 0 0 0
 1 1 1 0 1 1 1 0 0 1 0 1 0 0 1 1 0 0 1 0 1 1 1 0 1 0 1 1 1 1 0 1 1 1 1 1 1
 0 0 1 0 1 1 1 0 0 1 1 0 1 1 1 1 1 1 0 1 1 0 1 1 0 0 1 1 0 0 1 0 0 1 1 1 0
 1 1 0 0 0 0 1 1 1 1 1 0 0 1 1 1 1 1 1 1 1 1 0 1 1 1 1 1 0 1 1 1 1 1 0 1 1
 0 0 0 1 1 1 0 1 1 1 1 1 1 1 0 1 0 1 1 1 1 1 1 1 1 0 0 1 1 1 0 1 1 1 0 0 0
 0 0 1 1 1 1 1 1]
accuracy: 0.46441947565543074


c:\Users\chunl\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\base.py:457: UserWarning: X has feature names, but SGDClassifier was fitted without feature names
  warnings.warn(


array([[ 54, 109],
       [ 34,  70]], dtype=int64)

In [171]:
print("hinge - alpha= 0.0001\n")
print("accuracy:",accuracy_score(y_test,prediction))
print("\nhinge - alpha= 0.001\n")
print("accuracy:",accuracy_score(y_test,prediction6))
print("\nlog - alpha=0.0001\n")
print("accuracy:",accuracy_score(y_test,prediction2))
print("\nlog - alpha=0.001\n")
print("accuracy:",accuracy_score(y_test,prediction5))
print("\nhuber - alpha= 0.0001\n")
print("accuracy:",accuracy_score(y_test,prediction3))
print("\nhuber - alpha= 0.001\n")
print("accuracy:",accuracy_score(y_test,prediction4))



hinge - alpha= 0.0001

accuracy: 0.6067415730337079

hinge - alpha= 0.001

accuracy: 0.46441947565543074

log - alpha=0.0001

accuracy: 0.3895131086142322

log - alpha=0.001

accuracy: 0.5955056179775281

huber - alpha= 0.0001

accuracy: 0.48314606741573035

huber - alpha= 0.001

accuracy: 0.4157303370786517


hinge is a linear loss function(soft-margin) ->compare to exponential loss function , it might have a better performance at the beginning. But it will have a slower progress then exponential. It is difficult to optimize

log_loss is logistic loss function(exponential) -> that mean if there are bigger the data the lower the loss(higher accuracy) . In the observation log_loss is the most stable one

huber is smooth loss that bring tolerance to the outliers
It give less weight to large errors, and make it more robust to outliers

The higher the value alpha, the stronger the regularization
alpha become larger . Performance should be better, but there are exceptions


1st try
hinge - alpha= 0.0001
accuracy: 0.6104868913857678

hinge - alpha= 0.001
accuracy: 0.4344569288389513

log - alpha=0.0001
accuracy: 0.4344569288389513

log - alpha=0.001
accuracy: 0.6629213483146067

huber - alpha= 0.0001
accuracy: 0.4232209737827715

huber - alpha= 0.001
accuracy: 0.5842696629213483

2nd try
hinge - alpha= 0.0001
accuracy: 0.6104868913857678

hinge - alpha= 0.001
accuracy: 0.6629213483146067

log - alpha=0.0001
accuracy: 0.5767790262172284

log - alpha=0.001
accuracy: 0.6329588014981273

huber - alpha= 0.0001
accuracy: 0.4344569288389513

huber - alpha= 0.001
accuracy: 0.47191011235955055

3rd try

hinge - alpha= 0.0001
accuracy: 0.6067415730337079

hinge - alpha= 0.001
accuracy: 0.46441947565543074

log - alpha=0.0001
accuracy: 0.3895131086142322

log - alpha=0.001
accuracy: 0.5955056179775281

huber - alpha= 0.0001
accuracy: 0.48314606741573035

huber - alpha= 0.001
accuracy: 0.4157303370786517